command/agent_test.go

```diff
@@ -1691,3 +1695,304 @@ vault {
+			expectTemplateRender:      templateRendered(0),
+			templateErrorOnMissingKey: false,
+			expectExit:                true,
+			expectError:               false,
+		},
+		"true, with non-existent secret": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(true),
+			templateContents:          badTemplateContent,
+			expectTemplateRender:      "",
+			templateErrorOnMissingKey: false,
+			expectExit:                true,
+			expectError:               true,
+		},
+		"true, with missing key": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(true),
+			templateContents:          missingKeyTemplateContent,
+			expectTemplateRender:      missingKeyTemplateRender,
+			templateErrorOnMissingKey: false,
+			expectExit:                true,
+			expectError:               false,
+		},
+		"true, with missing key, with error_on_missing_key": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(true),
+			templateContents:          missingKeyTemplateContent,
+			expectTemplateRender:      "",
+			templateErrorOnMissingKey: true,
+			expectExit:                true,
+			expectError:               true,
+		},
+		"false, no template error": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(false),
+			templateContents:          templateContents(0),
+			expectTemplateRender:      templateRendered(0),
+			templateErrorOnMissingKey: false,
+			expectExit:                false,
+			expectError:               false,
+		},
+		"false, with non-existent secret": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(false),
+			templateContents:          badTemplateContent,
+			expectTemplateRender:      "",
+			templateErrorOnMissingKey: false,
+			expectExit:                false,
+			expectError:               true,
+		},
+		"false, with missing key": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(false),
+			templateContents:          missingKeyTemplateContent,
+			expectTemplateRender:      missingKeyTemplateRender,
+			templateErrorOnMissingKey: false,
+			expectExit:                false,
+			expectError:               false,
+		},
+		"false, with missing key, with error_on_missing_key": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(false),
+			templateContents:          missingKeyTemplateContent,
+			expectTemplateRender:      missingKeyTemplateRender,
+			templateErrorOnMissingKey: true,
+			expectExit:                false,
+			expectError:               true,
+		},
+		"missing": {
+			exitOnRetryFailure:        nil,
+			templateContents:          templateContents(0),
+			expectTemplateRender:      templateRendered(0),
+			templateErrorOnMissingKey: false,
+			expectExit:                false,
+			expectError:               false,
+		},
+	}
+
+	for tcName, tc := range testCases {
+		t.Run(tcName, func(t *testing.T) {
+			// create temp dir for this test run
+			tmpDir, err := ioutil.TempDir(tmpDirRoot, tcName)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			listenAddr := "127.0.0.1:18123"
```

Comment:
Fixed ports make tests sensitive to failures due to port clashes (even if with itself on a VM where two test suites are run in parallel for whatever reason). This case is a tricky one to do reliably, given we can't ask the kernel to reserve a random port for us in the way we normally would in tests, because the port is for use in a separate process. Could we reduce the chance of failure substantially by generating a random high port each time though?

Reply:
This bit of the code is borrowed from `TestAgent_Cache_Retry`, but as long as we're not parallelizing this test with any other tests (by explicitly calling `t.Parallel()`) it should be fine.